north/src/runtime/linux/cgroups.rs

```diff
@@ -23,181 +23,93 @@ use std::{
-                        return;
-                    }
-                };
-                time::sleep(Duration::from_millis(500)).await;
-            }
-        });
-
-        debug!("Created {}", path.display());
-        Ok(CGroupMem { path, monitor })
-    }
+    tokio::try_join!(
+        copy_file("cpuset.cpus", &mount_point, &cpuset_root),
+        copy_file("cpuset.mems", &mount_point, &cpuset_root),
+    )?;
+    Ok(())
 }
 
-#[derive(Debug)]
-pub struct CGroupCpu {
-    pub path: PathBuf,
+async fn copy_file(name: &str, src_dir: &Path, dst_dir: &Path) -> Result<(), LinuxError> {
+    let file_path = src_dir.join(name);
+    // NOTE For some reason using fs::copy here does not work
+    let value = fs::read_to_string(&file_path).await.map_err(|e| {
+        LinuxError::FileOperation(
+            format!("Could not read content of {}", file_path.display(),),
+            e,
+        )
+    })?;
+    write(&dst_dir.join(name), &value).await
 }
 
-impl CGroup for CGroupCpu {
-    fn path(&self) -> &Path {
-        self.path.as_path()
-    }
-}
+impl CGroup {
+    async fn new(
+        cgroup: String,
+        path: &Path,
+        config: &manifest::CGroup,
+    ) -> Result<CGroup, LinuxError> {
+        let path = get_mount_point(&cgroup)?.join(&path);
+        create_if_not_exists(&path).await?;
+
+        for (param, value) in config {
+            let filename = path.join(format!("{}.{}", &cgroup, param));
+            debug!("Setting {} to {}", filename.display(), value);
+            write(&filename, &value).await?;
+        }
 
-impl CGroupCpu {
-    pub async fn new(
-        parent: &Path,
-        name: &str,
-        cgroup: &manifest::CGroupCpu,
-    ) -> Result<CGroupCpu, LinuxError> {
-        let mount_point = get_mount_point("cpu")?;
-        let path = mount_point.join(parent).join(name);
-        create(&path).await?;
+        Ok(CGroup { cgroup, path })
+    }
 
-        // Configure cpu shares
-        let cpu_shares = path.join(CPU_SHARES);
-        debug!(
-            "Setting {} to {} shares",
-            cpu_shares.display(),
-            cgroup.shares
-        );
-        write(&cpu_shares, &cgroup.shares.to_string()).await?;
+    /// Assign a PID to this cgroup
+    async fn assign(&self, pid: u32) -> Result<(), LinuxError> {
+        debug!("Assigning {} to {}", pid, self.path.display());
+        let tasks = self.path.join(TASKS);
+        write(&tasks, &pid.to_string()).await
+    }
 
-        Ok(CGroupCpu { path })
+    /// Destroy the cgroup by removing the dir
+    async fn destroy(self) -> Result<(), LinuxError> {
+        debug!("Destroying cgroup {}", self.path.display());
+        fs::remove_dir(&self.path)
+            .await
+            .map_err(|e| Error::Destroy(self.path.to_string_lossy().to_string(), e).into())
     }
 }
 
 #[derive(Debug)]
-pub struct CGroups {
-    pub mem: Option<CGroupMem>,
-    pub cpu: Option<CGroupCpu>,
-}
+pub struct CGroups(Vec<CGroup>);
```

Comment:
In `config.rs` the type `CGroups` is a hashmap, here it is a vector...if we need different types they should have different names. But I guess we do not need them.

Reply:
There is a `cgroup` section in the `manifest.yml`, one in the `north.toml` config and finally the source `cgroups.rs`. Therefore the 3 definitions, related but not equivalent. I find this confusing too. I think we can't unify them all, as they represent slightly different things, maybe we can resort to just give them different names?

Reply:
A unified type would be nice if possible, but if not practical different types might be an option!